server/etcdserver/txn/txn.go

```diff
@@ -43,16 +41,23 @@ func Put(ctx context.Context, lg *zap.Logger, lessor lease.Lessor, kv mvcc.KV, t
 			traceutil.Field{Key: "req_size", Value: p.Size()},
 		)
 	}
-	val, leaseID := p.Value, lease.LeaseID(p.Lease)
-	if txnWrite == nil {
-		if leaseID != lease.NoLease {
-			if l := lessor.Lookup(leaseID); l == nil {
-				return nil, nil, lease.ErrLeaseNotFound
-			}
+	leaseID := lease.LeaseID(p.Lease)
+	if leaseID != lease.NoLease {
+		if l := lessor.Lookup(leaseID); l == nil {
+			return nil, nil, lease.ErrLeaseNotFound
 		}
-		txnWrite = kv.Write(trace)
-		defer txnWrite.End()
 	}
+	txnWrite := kv.Write(trace)
+	defer txnWrite.End()
+	resp, err = put(ctx, txnWrite, p)
+	return resp, trace, err
+}
+
+func put(ctx context.Context, txnWrite mvcc.TxnWrite, p *pb.PutRequest) (resp *pb.PutResponse, err error) {
+	trace := traceutil.Get(ctx)
```

Comment:
A separate trace as compared to the trace in `Put`? You may want to associate the trace with the ctx in `Put`

`ctx = context.WithValue(ctx, traceutil.TraceKey, trace)`

Reply:
Done